### PR TITLE
Add Move action

### PR DESCRIPTION
**Pull request by FabianDevel**

### CODE REVIEW

File: app/src/main/res/layout/fragment_move.xml [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+        android:layout_height="?attr/actionBarSize"
+        android:background="@color/background"
+        app:contentInsetEnd="0dp"
+        app:navigationIconTint="?android:attr/colorPrimary"
+        app:titleTextAppearance="@style/H2">
+
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:gravity="center_vertical">
```
Comment (LunarX) [inline app/src/main/res/layout/fragment_move.xml:37]: Even if fragment_thread had a constraint layout, it is actually not needed. A LinearLayout is enough

File: app/src/main/res/layout/fragment_move.xml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:animateLayoutChanges="true"
+            android:background="@color/backgroundColor"
+            android:scrollbars="none">
+
+            <LinearLayout
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:layout_marginVertical="@dimen/marginStandardMedium"
```
Comment (LunarX) [inline app/src/main/res/layout/fragment_move.xml:83]: Reduce margin top to 12dp or 8dp to better fit models

File: app/src/main/res/layout/item_menu_drawer.xml [hunk 3]
```diff
@@ -51,7 +54,22 @@
...
+            android:id="@+id/itemTick"
+            android:layout_width="wrap_content"
+            android:layout_height="16dp"
+            android:contentDescription="@string/contentDescriptionIconFolderSelected"
+            android:src="@drawable/ic_check"
+            android:visibility="gone"
+            app:layout_constraintBottom_toBottomOf="parent"
+            app:layout_constraintEnd_toEndOf="parent"
+            app:layout_constraintTop_toTopOf="parent"
+            app:tint="?android:attr/colorPrimary" />
```
Comment (LunarX) [inline app/src/main/res/layout/item_menu_drawer.xml:72]: Don't use a constraint layout, linear is fine enough. And use a FrameLayout to contain the bade and the checkmark if they can never be displayed together

**Review by LunarX — CHANGES_REQUESTED** [review]

File: app/src/main/res/layout/fragment_move.xml (hunk 2 above)
Comment (FabianDevel) [reply app/src/main/res/layout/fragment_move.xml:83]: 8dp seems a little bit to small, 12dp seems better

File: app/src/main/java/com/infomaniak/mail/views/MenuDrawerItemView.kt [hunk 4]
```diff
@@ -99,21 +101,27 @@ class MenuDrawerItemView @JvmOverloads constructor(
...
-        val (color, textAppearance) = if (isSelected) {
-            Pair(context.getAttributeColor(RMaterial.attr.colorPrimaryContainer), R.style.BodyMedium_Accent)
+        val (color, textAppearance) = if (isSelected && selectionStyle == SelectionStyle.MENU_DRAWER) {
+            context.getAttributeColor(RMaterial.attr.colorPrimaryContainer) to R.style.BodyMedium_Accent
         } else {
-            Pair(ContextCompat.getColor(context, R.color.backgroundColor), R.style.BodyMedium)
+            ContextCompat.getColor(context, R.color.backgroundColor) to R.style.BodyMedium
         }
 
+        itemTick.isVisible = isSelected && selectionStyle == SelectionStyle.MOVE_FRAGMENT
```
[the remark this answers is not in the document]
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/views/MenuDrawerItemView.kt:110]: done

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 5]
```diff
@@ -309,7 +306,7 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
         val thread = ThreadController.getThread(threadUid) ?: return null
 
         val messages = when (message) {
-            null -> MessageController.getArchivableMessages(thread, currentFolderId.value!!)
+            null -> MessageController.getMoveableMessages(thread, currentFolderId.value!!)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:309]:
```suggestion
            null -> MessageController.getMoveableMessages(thread, thread.folderId)
```

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 6]
```diff
@@ -419,6 +416,52 @@ class MainViewModel(application: Application) : AndroidViewModel(application) {
...
     //endregion
 
+    //region Move
+    fun moveTo(destinationFolderId: String, threadUid: String, messageUid: String? = null) {
+        viewModelScope.launch(Dispatchers.IO) {
+            val mailbox = currentMailbox.value ?: return@launch
+            val thread = ThreadController.getThread(threadUid) ?: return@launch
+            val message = messageUid?.let { MessageController.getMessage(messageUid) ?: return@launch }
+            val messages = when (message) {
+                null -> MessageController.getMoveableMessages(thread, currentFolderId.value!!)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:426]:
```suggestion
                null -> MessageController.getMoveableMessages(thread, thread.folderId)
```

File: app/src/main/java/com/infomaniak/mail/ui/main/settings/swipe/SwipeActionsSelectionSettingFragment.kt [hunk 7]
```diff
@@ -68,7 +68,7 @@ class SwipeActionsSelectionSettingFragment : Fragment() {
                 swipeAction as SwipeAction
 
                 when (swipeAction) {
-                    ARCHIVE, DELETE, FAVORITE, NONE, QUICKACTIONS_MENU, READ_UNREAD, READ_AND_ARCHIVE, SPAM -> Unit
+                    ARCHIVE, DELETE, FAVORITE, MOVE, NONE, QUICKACTIONS_MENU, READ_UNREAD, READ_AND_ARCHIVE, SPAM -> Unit
```
[the remark this answers is not in the document]
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/settings/swipe/SwipeActionsSelectionSettingFragment.kt:71]: I second this motion.

[the remark this answers is not in the document]
Comment (FabianDevel) [reply app/src/main/java/com/infomaniak/mail/ui/main/settings/swipe/SwipeActionsSelectionSettingFragment.kt:71]: The only other swipe action that is not implemented is postpone, which is not v1 so I thought about removing postpone in the layout and just delete this when as it doesn't have meaning.
But I wanted to do it in another PR

File: app/src/main/res/layout/item_menu_drawer.xml (hunk 3 above)
Comment (KevinBoulongne) [reply app/src/main/res/layout/item_menu_drawer.xml:72]: ⚔️

File: app/src/main/res/drawable/ic_folder_add.xml [hunk 8]
```diff
@@ -0,0 +1,30 @@
+<!--
```
Comment (KevinBoulongne) [inline app/src/main/res/drawable/ic_folder_add.xml:1]:
```suggestion
<?xml version="1.0" encoding="utf-8"?><!--
```

File: app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+        binding.toolbar.setNavigationOnClickListener { findNavController().popBackStack() }
+    }
+
+    private fun setupAdapters() = with(binding) {
+        defaultFoldersList.adapter = defaultFoldersAdapter
+        customFoldersList.adapter = customFoldersAdapter
+    }
+
+    private fun setupListeners() {
+        binding.inboxFolder.setOnClickListener { inboxFolderId?.let(::moveToFolder) }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt:71]:
```suggestion
        binding.inboxFolder.setOnClickListener { inboxFolderId?.let(::moveToFolder) }
        binding.toolbar.setNavigationOnClickListener { findNavController().popBackStack() }
```

File: app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+
+        setupAdapters()
+        setupListeners()
+        observeFolders()
+
+        binding.toolbar.setNavigationOnClickListener { findNavController().popBackStack() }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt:62]:
```suggestion
```

File: app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    private fun setupAdapters() = with(binding) {
+        defaultFoldersList.adapter = defaultFoldersAdapter
+        customFoldersList.adapter = customFoldersAdapter
+    }
+
+    private fun setupListeners() {
+        binding.inboxFolder.setOnClickListener { inboxFolderId?.let(::moveToFolder) }
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/menu/MoveFragment.kt:72]:
```suggestion
    private fun setupListeners() = with(binding) {
        inboxFolder.setOnClickListener { inboxFolderId?.let(::moveToFolder) }
        toolbar.setNavigationOnClickListener { findNavController().popBackStack() }
    }
```

**Review by KevinBoulongne — APPROVED** [review]

**Review by JorisBodin — APPROVED** [review]

**Pull request merged**